incubator/prisma/templates/configmap.yaml

```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+kind: ConfigMap
```

Comment:
@giacomoguiulfo I recommend converting this to secret.

Reply:
Thanks for the review!
I don't think the full config should be in a secret. But definitely the auth secret should be in a secret. I will work on this. I hope Prisma supports passing the auth secret in a different way.

Reply:
Yeah, this definitely need to be defined as a `Secret`. I'm actually surprised that the documentation does not mention this (https://www.prisma.io/docs/1.12/tutorials/deploy-prisma-servers/kubernetes-aiqu8ahgha).

In my chart I put the entire `PRISMA_CONFIG` into a secret.

That's the only way that you can currently use this (https://github.com/prismagraphql/prisma/blob/665eb16138dfea9082746e21743121eccfc41419/server/libs/prisma-config/src/main/scala/com/prisma/config/ConfigLoader.scala#L45-L52). It would make sense to be able to pass multiple ENV variables.

/cc @akoenig @nikolasburk

Reply:
Thanks for the review!
Yes, we could have the full config in a secret (for now). However, do you have any suggestion on how to approach this? The templating can get really uggly since it has to be base64 encoded and it is particularly long for a secret.

Reply:
That's pretty simple to do. I usually have a `secret.yaml` file that describes the `Secret` resource, reads up the secrets from the "values" and encodes them into the `Secret`.

```yaml
apiVersion: v1
kind: Secret
metadata:
  # The identifier of this Secret, unique within a namespace.
  # https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
  name: {{ .Chart.Name }}-secret
  # Labels are custom fields useful for querying and filtering resources.
  # https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
  labels:
    app: {{ .Chart.Name }}
    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    component: {{ .Values.componentName }}
    heritage: {{ .Release.Service }}
    release: {{ .Release.Name }}
type: Opaque
data:
  # Inject (sensitive) environment variables from `Values` config.
  {{- range $key, $val := .Values.secrets }}
  {{ $key }}: {{ $val | b64enc }}
  {{- end}}
```

This assumes that you have a `secrets` object map in the `values.yaml`.

```yaml
secrets:
  PRISMA_MANAGEMENT_API_SECRET: xxx
  PRISMA_CONFIG: |
    port: 4466
    managementApiSecret: xxx
    databases:
      default:
        connector: postgres
        host: localhost
        port: 5432
        user: postgres
        password: 123
        migrations: true
```

Because of the ability to "merge" values together, I usually have a `values.yaml` and a `secrets.yaml` which I pass to helm when installing

```bash
$ helm upgrade --install <release-name> \
  --namespace <namespace> \
  -f values/<env>/values.yaml \
  -f values/<env>/secrets.yaml \
  <chart-name>
```

Now, since the "secrets" should be encrypted, I use Google Cloud KMS to encrypt the `secrets.yaml` file into a `secrets.enc` and then decrypt the secrets on CI before upgrading the chart

```bash
$ gcloud kms decrypt \
  --project=<project-id> \
  --location=global \
  --keyring=<key-ring> \
  --key=<crypto-key> \
  --plaintext-file=values/<env>/secrets.yaml \
  --ciphertext-file=values/<enc>/secrets.enc
```

Finally, the secret values can be then referenced from the `Deployment`

```yaml
spec:
  template:
    spec:
      containers:
      - image: "prismagraphql/prisma:1.12"
        # Inject (non sensitive) environment variables from the `Values` config.
        env: {{- range $key, $val := .Values.config }}
        - name: {{ $key }}
          value: {{ $val | quote }}
          {{- end}}
        # Inject (sensitive) environment variables from `Values` config, referring
        # them from the related Secret object. This is an extra measure of security
        # to avoid having credentials displayed in plain text in the template file.
        # https://kubernetes.io/docs/concepts/configuration/secret/
        {{- $chartName := .Chart.Name -}}
        {{- range $key, $val := .Values.secrets }}
        - name: {{ $key }}
          valueFrom:
            secretKeyRef:
              name: {{ $chartName }}-secret
              key: {{ $key }}
          {{- end}}
```

Reply:
Looks good 👍 I will add it soon. Thanks!

Reply:
@emmenko Actually, it's not possible through this method because `host: localhost` is a value that depends on the release name, so it can't be something in the `values.yaml` file AFAIK.